### PR TITLE
[API] Initial API tests

### PR DESCRIPTION
**Pull request by lchrusciel**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #3468
| License       | MIT
| Doc PR        | [#371](https://github.com/Sylius/Sylius-Docs/pull/372)

I want to introduce initial tests for Sylius API. The previous ApiTestCase was removed from Sylius and after small refactoring and a few improvements we prepared a stand-alone library [Lakion ApiTestCase](https://github.com/Lakion/ApiTestCase). I added this library to composer and wrote some sample test for Shipping Categories. I want to test all available endpoints so feel free to help me with this task.

*Update:*
**NOTE:** Because of depending on [PHP to string](https://github.com/coduo/php-to-string), which has a class String the tests will fail on PHP7. We are waitng to merge [PR](https://github.com/coduo/php-to-string/pull/3) which will resolve this issue.

*Update2:*
**NOTE:** [PHP to string](https://github.com/coduo/php-to-string) was fixed and it supports PHP7. But because we are still waiting for the official release temporary I added "minimum-stability": "dev" to composer. Do not merge it yet.  

*Update3:*
**NOTE:** As far as we are good with depending on PHP-Matcher dev explicitly its ready for merge. At the end, ApiTestCase doesn't have a stable version to.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php [hunk 1]
```diff
@@ -0,0 +1,207 @@
+<?php
+
+namespace Sylius\Bundle\CoreBundle\Tests\Controller;
+
+use Lakion\ApiTestCase\JsonApiTestCase;
+
+/**
+ * @author Łukasz Chruściel <lukasz.chrusciel@lakion.com>
+ */
+class ShippingCategoryControllerTest extends JsonApiTestCase
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: I am wondering if we should name tests like this. The ShippingCategoryController class does not exist and people are used to the fact that there is usually Xyz class and XyzTest. Something that sounds less like a class will be less confusing I hope.

``ShippingCategoryApiTest`` vs ``ShippingCategoryControllerTest``

WDYT?

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    {
+        $this->client->request('POST', '/api/shipping-categories/');
+
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'authentication/access_denied', 401);
+    }
+
+    public function testCreateShippingCategoryValidationFailResponse()
+    {
+        $this->resetDataBase();
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: ``resetDatabase`` or ``purgeDatabase``. Why we renamed it to reset?

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'shipping_category/new_shipping_category', 201);
+    }
+
+    public function testGetShippingCategoriesListAccessDeniedResponse()
+    {
+        $this->client->request('GET', '/api/shipping-categories/');
+
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'authentication/access_denied', 401);
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: I am wondering if we should test AccessDenied for every path of the API. I know it is important, but we do not do that in Behat and it adds a lot of redundant tests. I think we could have one TestCase that checks security on few endpoints and that would be more reasonable approach. Economy of Tests is important.

Thoughts?

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    public function testCreateShippingCategoryResponse()
+    {
+        $this->resetDataBase();
+
+        $this->client->request('POST', '/api/shipping-categories/', [], [], [
+            'HTTP_Authorization' => 'Bearer access_token',
+            'CONTENT_TYPE' => 'application/json',
+        ], '{"name": "Regular", "description": "Regular weight items", "code": "reg"}');
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: Can we somehow make this call and JSON more readable here? Can we write JSON in "pretty" form?

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+     * @return integer
+     */
+    private function createShippingCategory()
+    {
+        $this->client->request('POST', '/api/shipping-categories/', [], [], [
+            'HTTP_Authorization' => 'Bearer access_token',
+            'CONTENT_TYPE' => 'application/json',
+        ], '{"name": "Regular", "description": "Regular weight items", "code": "reg"}');
+
+        $response = $this->client->getResponse();
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: Missing blank line before return statement.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        ]);
+
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'shipping_category/updated_shipping_category', 200);
+    }
+
+    private function resetDataBase()
+    {
+        $this->purgeDataBase();
+        $this->loadPlatformData();
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: ``loadPlatformData`` is a weird name. Why not ``purgeDatabase`` and ``loadFixtures``?

File: src/Sylius/Bundle/CoreBundle/Tests/DataFixtures/ORM/api_administrator.yml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+Sylius\Bundle\ApiBundle\Model\AccessToken:
+    oauth_token:
+        client: @oauth
+        user: @admin
+        token: access_token
+
+Sylius\Component\Core\Model\User:
+    admin:
+        plainPassword: sylius
+        roles: [ROLE_API]
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/DataFixtures/ORM/api_administrator.yml]: I think we use ``ROLE_API_ACCESS``.

File: src/Sylius/Bundle/CoreBundle/Tests/DataFixtures/ORM/api_administrator.yml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    oauth:
+        randomId: client_id
+        secret: secret
+        allowedGrantTypes: [password, access_token, refresh_token]
+
+Sylius\Bundle\ApiBundle\Model\AccessToken:
+    oauth_token:
+        client: @oauth
+        user: @admin
+        token: access_token
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/DataFixtures/ORM/api_administrator.yml]: Maybe we can use sample token that looks like a real token? Makes tests nicer.

File: src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/authentication/access_denied.json [hunk 9]
```diff
@@ -0,0 +1,4 @@
+{
+    "error": "access_denied",
+    "error_description": "OAuth2 authentication required"
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/authentication/access_denied.json]: Kinda sucks that we have different format for most errors and 403 but we can take care of that later.

File: src/Sylius/Bundle/ShippingBundle/Resources/config/serializer/Model.ShippingCategory.yml [hunk 10]
```diff
@@ -6,24 +6,32 @@ Sylius\Component\Shipping\Model\ShippingCategory:
...
         - rel: self
           href:
-                route: sylius_api_shipping_category_show
-                parameters:
-                    id: expr(object.getId())
+            route: sylius_api_shipping_category_show
+            parameters:
+                id: expr(object.getId())
+            exclusion:
+                groups: [Details]
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ShippingBundle/Resources/config/serializer/Model.ShippingCategory.yml]: To be honest I do not know what is the recommendation here. Have you done some research?

**Comment by pjedrzejewski** [comment]
Nice work Łukasz! Let's hear more comments.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php (hunk 3 above)
Comment (amenophis) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: Maybe do an automated test through a loop on all api routes ?

File: src/Sylius/Bundle/CoreBundle/Tests/DataFixtures/ORM/api_administrator.yml (hunk 7 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Tests/DataFixtures/ORM/api_administrator.yml]: According to data fixtures we use [`ROLE_API`](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/FixturesBundle/DataFixtures/ORM/LoadApiData.php#L38)

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php (hunk 6 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: It is changed in the newest version of ApiTestCase. Whole this function will be removed.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php (hunk 1 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: I'm for `ShippingCategoryApiTest` to minimalize confusion.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        $this->client->request('POST', '/api/shipping-categories/', [], [], [
+            'HTTP_Authorization' => 'Bearer access_token',
+            'CONTENT_TYPE' => 'application/json',
+        ]);
+
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'shipping_category/create_validation_fail', 400);
+    }
+
+    public function testCreateShippingCategoryResponse()
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: I'm wondering if we should suffix all methods with `Response`. Isn't it enough to call it `testCreateShippingCategory`?

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php (hunk 1 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: `ShippingCategoryApiTest` is fine for me.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php (hunk 11 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: The question is: do we test logic or received response?

File: src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/shipping_category/shipping_category.json [hunk 12]
```diff
@@ -0,0 +1,8 @@
+{
+    "id": @integer@,
+    "name": "Regular",
+    "code": "reg",
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/shipping_category/shipping_category.json]: We should stick to one order of fields in a response. IMO `id`, `code`, `name`, etc. as below.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php (hunk 11 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: We are testing response contents/headers only so I am fine with leaving that suffix.

File: src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/shipping_category/shipping_category.json (hunk 12 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/shipping_category/shipping_category.json]: :+1:

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php (hunk 11 above)
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: IMHO we test logic and we check if it's correct by comparing received responses.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php (hunk 3 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: These tests are really fast(we don't even load fixtures for this), so I think it is safer to have them. Maybe we can move them to separate TestCase.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+        $this->client->request('GET', '/api/shipping-categories/'.$id, [], [], [
+            'HTTP_Authorization' => 'Bearer access_token',
+            'ACCEPT' => 'application/json',
+        ]);
+
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'shipping_category/updated_shipping_category', 200);
+    }
+
+    private function resetDataBase()
```
Comment (mheki) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryControllerTest.php]: `DataBase` with capital `B` looks odd IMO ;)

File: phpunit.xml.dist [hunk 14]
```diff
@@ -14,11 +14,9 @@
...
 
-    <!--
     <php>
-        <server name="KERNEL_DIR" value="/path/to/your/app/" />
+        <server name="KERNEL_DIR" value="/app/" />
     </php>
-    -->
 
     <filter>
         <whitelist>
```
Comment (bendavies) [inline phpunit.xml.dist]: the paths in the rest if the file are wrong. it lists `../src` intead of just `src`

File: src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/error/not_found.json [hunk 15]
```diff
@@ -0,0 +1,4 @@
+{
+    "code": 404,
+    "message": @string@
```
Comment (bendavies) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/error/not_found.json]: you can actually enclose these placeholders in `"`. It still works.
This has the added benefit of being valid json, so can be interpreted/formatted by an IDE.

File: src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/shipping_category/shipping_category.json [hunk 16]
```diff
@@ -0,0 +1,8 @@
...
+    "description": "Regular weight items",
+    "created_at": @string@,
```
Comment (bendavies) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/shipping_category/shipping_category.json]: for dates you can do `"@string@.isDateTime()",`

File: travis/suites/suite-default [hunk 17]
```diff
@@ -1,5 +1,6 @@
 #!/bin/bash
 bin/phpspec run -f dot
+bin/phpunit src/ -c phpunit.xml.dist
```
Comment (bendavies) [inline travis/suites/suite-default]: as you've moved `phpunit.xml.dist` to the root, you should just be able to run `bin/phpunit` here.

File: src/Sylius/Bundle/ShippingBundle/Resources/config/serializer/Model.ShippingCategory.yml (hunk 10 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/ShippingBundle/Resources/config/serializer/Model.ShippingCategory.yml]: Not yet. I just though that there is no point to have link to show, if you have already detailed information.

File: src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/shipping_category/shipping_category.json (hunk 16 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/shipping_category/shipping_category.json]: :+1:

File: src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/error/not_found.json (hunk 15 above)
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/error/not_found.json]: :+1:

Comment (bendavies) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/error/not_found.json]: just thought, probably can't do this for `@integer@`.

**Comment by lchrusciel** [comment]
@bendavies Thanks for advices! I have to update ApiTestCase README to, to point out these features ;)

File: src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/error/not_found.json (hunk 15 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/error/not_found.json]: it works!

Comment (bendavies) [inline src/Sylius/Bundle/CoreBundle/Tests/Responses/Expected/error/not_found.json]: nice

File: phpunit.xml.dist [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+>
+    <testsuites>
+        <testsuite name="Project Test Suite">
+            <directory>src/*/*Bundle/Tests</directory>
+            <directory>src/*/Bundle/*Bundle/Tests</directory>
+        </testsuite>
+    </testsuites>
+
+    <php>
+        <server name="KERNEL_CLASS_PATH" value="/app/TestKernel.php" />
```
Comment (bendavies) [inline phpunit.xml.dist]: missing `KERNEL_DIR`?

Comment (lchrusciel) [inline phpunit.xml.dist]: ApiTestCase allows you to use Kernel class directly. We need it, because we have two separated Kernels. The problem is with version of ApiTestCase on travis.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+            'CONTENT_TYPE' => 'application/json',
+        ]);
+
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'shipping_category/create_validation_fail', 400);
+    }
+
+    public function testCreateShippingCategoryResponse()
+    {
+        $this->loadFixturesFromFile('api_administrator.yml');
```
Comment (Arminek) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php]: This method has been changed in api test case.

File: composer.json [hunk 20]
```diff
@@ -203,5 +203,6 @@
         },
         "symfony-app-dir": "app",
         "symfony-web-dir": "web"
-    }
+    },
+    "minimum-stability": "dev"
```
Comment (bendavies) [inline composer.json]: i don't think you should do this, you should require `"coduo/php-matcher": "2.1.*@dev",` explicitly

**Comment by aRn0D** [comment]
@pjedrzejewski `WebTestCase` does not send a real http request, in some case it hides some problems. Behat is really cool for that kind of things but there is not a very good extension to manage REST.

File: phpunit.xml.dist [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+<!-- http://phpunit.de/manual/4.1/en/appendixes.configuration.html -->
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/4.1/phpunit.xsd"
+         backupGlobals="false"
+         colors="true"
+         bootstrap="app/bootstrap.php.cache"
+>
+    <testsuites>
+        <testsuite name="Project Test Suite">
```
Comment (pjedrzejewski) [inline phpunit.xml.dist]: Sylius Test Suite

**Comment by pamil** [comment]
@aRn0D neither *WebTestCase* nor our Behat with *Symfony2Extension* (without JS) send a real HTTP request as both of them use `Symfony\Bundle\FrameworkBundle\Client`

File: phpunit.xml.dist [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+            <directory>src/*/*Bundle/Tests</directory>
```
Comment (pamil) [inline phpunit.xml.dist]: Do we have any directory that matches this pattern? Shouldn't it be `src/*/Component/*/Tests` as the component equivalent of the bundle one line below?

**Comment by aRn0D** [comment]
@pamil I know but it is enough for web case (get and post) when you use put method and you want to update a file, it is another story ;). PHP does not decode HTTP header... I just wanted to warn you but anyway, even if I prefer behat I prefer you test the API with any solution because : No test no beer

**Comment by pamil** [comment]
@aRn0D didn't know about that, thanks! Does it happen only on PUT request, but not on POST ones? Last week I was uploading image in some scenario on while creating taxonomy (POST) and it worked like a charm.

File: src/Sylius/Bundle/ApiBundle/Resources/config/routing/shipping_category.yml [hunk 23]
```diff
@@ -7,6 +7,7 @@ sylius_api_shipping_category_index:
     defaults:
         _controller: sylius.controller.shipping_category:indexAction
         _sylius:
+            serialization_groups: [Default]
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ApiBundle/Resources/config/routing/shipping_category.yml]: Is it needed? Isn't `Default` a default serialization group?

**Comment by aRn0D** [comment]
@pamil Just with PUT and PATCH I did not test the latest version of PHP.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+
+        $this->client->request('POST', '/api/shipping-categories/', [], [], [
+            'HTTP_Authorization' => 'Bearer SampleTokenNjZkNjY2MDEwMTAzMDkxMGE0OTlhYzU3NzYyMTE0ZGQ3ODcyMDAwM2EwMDZjNDI5NDlhMDdlMQ',
+            'CONTENT_TYPE' => 'application/json',
+        ]);
+
+        $response = $this->client->getResponse();
+        $this->assertResponse($response, 'shipping_category/create_validation_fail', 400);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php]: Shouldn't we use constant for codes? E.g. `Response::HTTP_BAD_REQUEST` here.

Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php]: Yes, we should ;)

File: composer.json [hunk 25]
```diff
@@ -92,8 +92,9 @@
         "behat/mink-selenium2-driver": "~1.2",
         "behat/mink": "~1.6",
         "akeneo/phpspec-skip-example-extension": "~1.2",
-        "coduo/php-matcher": "~1.0",
         "lakion/mink-debug-extension": "^1.0.1",
+        "lakion/api-test-case": "dev-master",
```
Comment (pjedrzejewski) [inline composer.json]: We should set a branch alias on ApiTestCase's composer.json and require here: ^1.0@dev.

File: phpunit.xml.dist (hunk 22 above)
Comment (pjedrzejewski) [inline phpunit.xml.dist]: I'd remove this. Components don't have PHPUnit test and won't have them anytime soon I guess.

File: src/Sylius/Bundle/ApiBundle/Resources/config/routing/shipping_category.yml (hunk 23 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ApiBundle/Resources/config/routing/shipping_category.yml]: Right now:

https://github.com/Sylius/SyliusResourceBundle/blob/master/Controller/Configuration.php#L369

And I think JMS requires to specify Default explicitly as it won't add it automatically if you pass an empty array - not sure. So I think we should add 'Default' to the Configuration::getSerializationGroups() method which I linked above.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php (hunk 24 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php]: :+1:

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    public function testCreateShippingCategoryResponse()
+    {
+        $this->loadFixturesFromFile('api_administrator.yml');
+
+        $this->client->request('POST', '/api/shipping-categories/', [], [], [
+            'HTTP_Authorization' => 'Bearer SampleTokenNjZkNjY2MDEwMTAzMDkxMGE0OTlhYzU3NzYyMTE0ZGQ3ODcyMDAwM2EwMDZjNDI5NDlhMDdlMQ',
+            'CONTENT_TYPE' => 'application/json',
+        ], '{"name": "Regular", "description": "Regular weight items", "code": "reg"}');
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php]: https://github.com/lchrusciel/Sylius/pull/1

Also, ``code`` should always be first, then name, description etc.

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php [hunk 27]
```diff
@@ -0,0 +1,207 @@
...
+     * @return integer
+     */
+    private function createShippingCategory()
+    {
+        $this->client->request('POST', '/api/shipping-categories/', [], [], [
+            'HTTP_Authorization' => 'Bearer SampleTokenNjZkNjY2MDEwMTAzMDkxMGE0OTlhYzU3NzYyMTE0ZGQ3ODcyMDAwM2EwMDZjNDI5NDlhMDdlMQ',
+            'CONTENT_TYPE' => 'application/json',
+        ], '{"name": "Regular", "description": "Regular weight items", "code": "reg"}');
+
+        $response = $this->client->getResponse();
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php]: Missing blank line before use statement.

File: src/Sylius/Bundle/CoreBundle/composer.json [hunk 28]
```diff
@@ -63,7 +63,8 @@
     "require-dev": {
         "phpspec/phpspec": "~2.1",
         "fzaninotto/faker": "~1.2",
-        "symfony/form": "^2.7"
+        "symfony/form": "^2.7",
+        "lakion/api-test-case": "dev-master"
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/composer.json]: Same comment as for main composer.json.

**Comment by pjedrzejewski** [comment]
Please change the payload formatting according to my sample PR if you agree and update ApiTestCase version. And we are good to go! Good job Łukasz!

File: src/Sylius/Bundle/ApiBundle/Resources/config/routing/shipping_category.yml (hunk 23 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/routing/shipping_category.yml]: Shouldn't it be done in a separate PR?

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php (hunk 27 above)
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php]: Almost. Before `return` statement.

Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php]: LOL, yeah. :D

**Comment by pjedrzejewski** [comment]
@lchrusciel Please rebase. :)

**Comment by lchrusciel** [comment]
@pjedrzejewski done! I have done a little refactoring also, cuz of changes in ApiTestCase and feature added by @bendavies.

**Comment by pjedrzejewski** [comment]
Awesome work, thank you Łukasz! :+1:

File: src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php [hunk 29]
```diff
@@ -0,0 +1,203 @@
...
+
+    public function testPartialUpdateShippingCategoryResponse()
+    {
+        $this->loadFixturesFromFile('authentication/api_administrator.yml');
+        $shippingCategories = $this->loadFixturesFromFile('resources/shipping_categories.yml');
+
+        $this->client->request('PATCH', '/api/shipping-categories/'.$shippingCategories['shipping_category_1']->getId(), [], [], [
+            'HTTP_Authorization' => 'Bearer SampleTokenNjZkNjY2MDEwMTAzMDkxMGE0OTlhYzU3NzYyMTE0ZGQ3ODcyMDAwM2EwMDZjNDI5NDlhMDdlMQ',
+            'CONTENT_TYPE' => 'application/json',
+        ], '{"name": "Light", "description": "Light weight items"}');
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Tests/Controller/ShippingCategoryApiTest.php]: Would be great to update these to use the format I suggested above.

**Comment by pjedrzejewski** [comment]
@lchrusciel Please have a look at my comment above and also... seems like we missed test for delete? /cc @michalmarcinkowski
